### PR TITLE
Factor out the code that checks for a PB client connecting to an HTTP server hanging.

### PR DESCRIPTION
**Pull request by tomprince**
`buildbot_worker.pbutil` contains some code to detect if the worker connects to the HTTP port of the master. Factor that code into a wrapper that doesn't use inheritance, in preparation for supporting connecting via endpoints. (This is part of spitting up and improving #2387).

### CODE REVIEW

File: worker/buildbot_worker/pbutil.py [hunk 1]
```diff
@@ -68,15 +63,18 @@ def clientConnectionLost(self, connector, reason):
...
     def clientConnectionMade(self, broker):
         self.resetDelay()
         PBClientFactory.clientConnectionMade(self, broker)
         self.doLogin(self._root, broker)
         self.gotRootObject(self._root)
 
+    def buildProtocol(self, addr):
+        return PBClientFactory.buildProtocol(self, addr)
+
+    def _hung_connection(self):
```
Comment (seankelly) [inline worker/buildbot_worker/pbutil.py]: Is this method used? I see a reference to `_hung_connection` in `Worker` but it has a `_hung_connection` method.

Comment (tomprince) [inline worker/buildbot_worker/pbutil.py]: It was used in a previous iteration, but can now be removed.

File: worker/buildbot_worker/test/test_util_hangcheck.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        result = Deferred()
+
+        site = Site(Data("", "text/plain"))
+        client = HangCheckFactory(
+            PBClientFactory(), lambda: result.callback(None))
+
+        self.patch(HangCheckProtocol, '_HUNG_CONNECTION_TIMEOUT', 0.1)
+
+        listen(self, TCP4ServerEndpoint(
+            reactor, 0), site).addCallback(
```
Comment (tardyp) [inline worker/buildbot_worker/test/test_util_hangcheck.py]: nodeJS/lisp style. I'm lost :)

I would avoid to make a multiline lambda, and make a proper def

    server_d = listen(self, TCP4ServerEndpoint(reactor, 0), site)
    @server_d.addCallback
    def buildClient(listening_port):
        TCP4ClientEndpoint(reactor, '127.0.0.1', listening_port.getHost().port).connect(client)

Comment (tomprince) [inline worker/buildbot_worker/test/test_util_hangcheck.py]: I went further and factored this entire logic into a separate helper.

**Comment by tardyp** [comment]
thanks!

**Pull request merged**